### PR TITLE
[ISSUES-3623] Project management encrypts user warehouse passwords

### LINKED ISSUE #3623: [Feature] Project management encrypts user warehouse passwords
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-streampark/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

In the project management module, if the private warehouse is selected, the password is not encrypted during the database storage and interface transmission after the user enters the password information, which may cause the risk of password leakage for enterprise users

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by jianjun159**
## What changes were proposed in this pull request

Issue Number: close #3623  

## Brief change log
Encrypts and salts private warehouse authentication passwords

## Verifying this change

This time the changes are minor

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): (yes / no)
no

### CODE REVIEW

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java [hunk 1]
```diff
@@ -116,6 +132,38 @@ public boolean update(Project projectParam) {
...
+            log.error("project password encrypt failed");
+            throw new ApiAlertException(e);
+          }
+        }
+      }
+    }
+    project.setPom(projectParam.getPom());
+    project.setDescription(projectParam.getDescription());
+    project.setBuildArgs(projectParam.getBuildArgs());
     updateInternal(projectParam, project);
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java]: The code in  135 ~ 139,164 ~ 166 is the same as the logic in the  updateInternal method, we need delete lines 135 ~ 139,164 ~ 166

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -94,7 +100,17 @@ public RestResponse create(Project project) {
     RestResponse response = RestResponse.success();
 
     ApiAlertException.throwIfTrue(count > 0, "project name already exists, add project failed");
-
+    if (StringUtils.isNotBlank(project.getPassword())) {
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java:103]: We can make the code more cleaner:

```
@Override
  public boolean update(Project projectParam) {
    Project project = getById(projectParam.getId());
    AssertUtils.notNull(project);
    ApiAlertException.throwIfFalse(
        project.getTeamId().equals(projectParam.getTeamId()),
        "TeamId can't be changed, update project failed.");
    ApiAlertException.throwIfFalse(
        !project.getBuildState().equals(BuildStateEnum.BUILDING.get()),
        "The project is being built, update project failed.");

    updateInternal(projectParam, project);

    if (project.isHttpRepositoryUrl()) {
      if (StringUtils.isBlank(projectParam.getUserName())) {
        project.setUserName(null);
        project.setPassword(null);
        project.setSalt(null);
      } else {
        project.setUserName(projectParam.getUserName());
        if (!Objects.equals(projectParam.getPassword(), project.getPassword())) {
          try {
            String salt = ShaHashUtils.getRandomSalt();
            String encrypt = EncryptUtils.encrypt(projectParam.getPassword(), salt);
            project.setPassword(encrypt);
            project.setSalt(salt);
          } catch (Exception e) {
            log.error("project password encrypt failed");
            throw new ApiAlertException(e);
          }
        }
      }
    }

    if (projectParam.getBuildState() != null) {
      project.setBuildState(projectParam.getBuildState());
      if (BuildStateEnum.NEED_REBUILD == BuildStateEnum.of(projectParam.getBuildState())) {
        List<Application> applications = listApps(project);
        // Update deployment status
        applications.forEach(
            (app) -> {
              log.info(
                  "update deploy by project: {}, appName:{}", project.getName(), app.getJobName());
              app.setRelease(ReleaseStateEnum.NEED_CHECK.get());
              applicationManageService.updateRelease(app);
            });
      }
    }
    baseMapper.updateById(project);
    return true;
  }

  private static void updateInternal(Project projectParam, Project project) {
    project.setName(projectParam.getName());
    project.setUrl(projectParam.getUrl());
    project.setBranches(projectParam.getBranches());
    project.setPrvkeyPath(projectParam.getPrvkeyPath());
    project.setPom(projectParam.getPom());
    project.setDescription(projectParam.getDescription());
    project.setBuildArgs(projectParam.getBuildArgs());
  }
```

File: streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql [hunk 3]
```diff
@@ -191,6 +191,7 @@ create table `t_flink_project` (
   `user_name` varchar(64) collate utf8mb4_general_ci default null,
   `password` varchar(64) collate utf8mb4_general_ci default null,
   `prvkey_path` varchar(128) collate utf8mb4_general_ci default null,
+  `salt` varchar(26) collate utf8mb4_general_ci default null,
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql:194]: Why is the length 26? I don’t know if it is too short, can you confirm it?

Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql:194]: We need to add this field in the [upgrade script](https://github.com/apache/incubator-streampark/blob/dev/streampark-console/streampark-console-service/src/main/assembly/script/upgrade/mysql/2.2.0.sql)

**Comment by wolfboys** [comment]
You can execute the command to fix the checkstyle issue:
mvn spotless:apply

File: streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql (hunk 3 above)
Comment (jianjun159) [reply streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql:194]: The salt length is used in the ShaHashUtils salt generation method, the default length is 26, but the password will be longer after encryption, resulting in database writing problems

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java (hunk 1 above)
Comment (jianjun159) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ProjectServiceImpl.java]: The code has been optimized in a more concise direction

File: streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql [hunk 4]
```diff
@@ -189,8 +189,9 @@ create table `t_flink_project` (
   `url` varchar(255) collate utf8mb4_general_ci default null,
   `branches` varchar(64) collate utf8mb4_general_ci default null,
   `user_name` varchar(64) collate utf8mb4_general_ci default null,
-  `password` varchar(64) collate utf8mb4_general_ci default null,
+  `password` varchar(512) collate utf8mb4_general_ci default null,
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql:192]: 512 😳, I think this length is too long? Can you confirm it?

Comment (jianjun159) [reply streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql:192]: This situation is in response to very long passwords, and it cannot be ruled out that there are users whose passwords are very long

**Comment by wolfboys** [comment]
Can you provide a testcase for encryption and decryption of passwords？

**Comment by jianjun159** [comment]
OK, I'll work on the encryption and decryption

benjobs ***@***.***> 于2024年3月27日周三 17:04写道：

**Comment by jianjun159** [comment]
> Can you provide a testcase for encryption and decryption of passwords？
The encryption and decryption method used in this change uses the corresponding method in EncryotUtils that has been encapsulated in the system, and the corresponding EncryptUtilsTest already exists in the test class, mainly for the security logic of the system itself

**Review by wolfboys — APPROVED** [review]
LGTM

**Pull request merged**